lib/ssz_ex.ex

```diff
@@ -367,26 +377,67 @@ defmodule LambdaEthereumConsensus.SszEx do
     end
   end
 
-  defp decode_bitvector(bit_vector, size) when bit_size(bit_vector) == size,
-    do: {:ok, BitVector.new(bit_vector, size)}
+  defp decode_bitlist(_bit_list, _max_size), do: {:error, "invalid bitlist"}
+
+  defp decode_bitvector(bit_vector, size) do
+    num_bytes = byte_size(bit_vector)
+
+    cond do
+      bit_size(bit_vector) == 0 ->
+        {:error, "ExcessBits"}
+
+      num_bytes != max(1, div(size + 7, 8)) ->
+        {:error, "InvalidByteCount"}
 
-  defp decode_bitvector(_bit_vector, _size), do: {:error, "invalid bit_vector length"}
+      true ->
+        case bit_vector do
+          # Padding bits are clear
+          <<_first::binary-size(num_bytes - 1), 0::size(8 - rem(size, 8) &&& 7),
+            _rest::bitstring>> ->
+            {:ok, BitVector.new(bit_vector, size)}
+
+          _else ->
+            {:error, "ExcessBits"}
+        end
+    end
+  end
 
-  defp decode_list(binary, basic_type, size) do
+  defp decode_list(binary, basic_type, size, schema) do
     fixed_size = get_fixed_size(basic_type)
 
-    with {:ok, decoded_list} = result <-
-           binary
-           |> decode_chunk(fixed_size, basic_type)
-           |> flatten_results() do
-      if length(decoded_list) > size do
-        {:error, "invalid max_size of list"}
-      else
+    if check_valid_list_size_prev_decode(schema, fixed_size, size, binary) do
+      with {:ok, decoded_list} = result <-
+             binary
+             |> decode_chunk(fixed_size, basic_type)
+             |> flatten_results(),
+           :ok <- check_valid_list_size(schema, size, length(decoded_list)) do
         result
       end
+    else
+      {:error, "invalid list binary size"}
     end
   end
 
+  def check_valid_list_size_prev_decode({:vector, _, _}, fixed_size, size, binary)
+      when fixed_size * size == byte_size(binary),
+      do: true
+
+  def check_valid_list_size_prev_decode({:vector, _, _}, _fixed_size, _size, _binary), do: false
+
+  def check_valid_list_size_prev_decode({:list, _, _}, _fixed_size, _size, _binary), do: true
```

Comment:
The fact that list don't have a pre-check is a big evidence towards this being over-reutilized.